### PR TITLE
Add unit tests for untested modules and update PR testing requirements

### LINKED ISSUE #4382: Add comprehensive test coverage for untested modules
## Summary

Multiple core modules in the codebase had **zero test coverage**. This issue tracks adding unit tests for all previously untested modules and updating project documentation to enforce testing as a PR requirement.

## Problem

After auditing the codebase, the following modules were found to have **no tests at all**:

- `accounting/` — models (LineItemType, Account, Transfer) and controllers (GlobalAccountingController, ProgramAccountingController, IndividualAccountingController)
- `application/` — StudentProgramApp, StudentClassApp models
- [cal/](cci:1://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/tests/test_context_processors.py:39:4-47:37) — Event and EventType models
- [context_processors.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/context_processors.py:0:0-0:0) — all 8 context processor functions
- `db/` — AjaxForeignKey custom field
- `dbmail/` — send_mail, ActionHandler, MessageRequest, cronmail
- [formstack/](cci:1://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/application/models.py:56:4-60:38) — FormstackForm, FormstackSubmission objects
- `middleware/` — ESPAuthMiddleware, ESPErrorMiddleware
- `miniblog/` — AnnouncementLink, Entry, Comment models
- `program/controllers/` — ConfirmationEmailController, ClassCreationController
- `survey/` — Survey, Question, Answer, ListField models
- [varnish/](cci:1://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/varnish/varnish.py:5:0-11:19) — cache purge utilities

### New Test Files (16 files, ~140 test methods)

All placed in `esp/esp/tests/` with filenames indicating the source file under test:

| Test File | Source Module |
|-----------|--------------|
| [test_accounting_models.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/tests/test_accounting_models.py:0:0-0:0) | [accounting/models.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/accounting/models.py:0:0-0:0) |
| [test_accounting_controllers.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/tests/test_accounting_controllers.py:0:0-0:0) | [accounting/controllers.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/accounting/controllers.py:0:0-0:0) |
| [test_application_models.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/tests/test_application_models.py:0:0-0:0) | [application/models.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/application/models.py:0:0-0:0) |
| [test_cal_models.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/tests/test_cal_models.py:0:0-0:0) | [cal/models.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/cal/models.py:0:0-0:0) |
| [test_classreg_controller.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/tests/test_classreg_controller.py:0:0-0:0) | [program/controllers/classreg.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/program/controllers/classreg.py:0:0-0:0) |
| [test_context_processors.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/tests/test_context_processors.py:0:0-0:0) | [context_processors.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/context_processors.py:0:0-0:0) |
| [test_db_fields.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/tests/test_db_fields.py:0:0-0:0) | [db/fields.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/db/fields.py:0:0-0:0) |
| [test_dbmail_cronmail.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/tests/test_dbmail_cronmail.py:0:0-0:0) | [dbmail/cronmail.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/dbmail/cronmail.py:0:0-0:0) |
| [test_dbmail_models.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/tests/test_dbmail_models.py:0:0-0:0) | [dbmail/models.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/dbmail/models.py:0:0-0:0) |
| [test_formstack_objects.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/tests/test_formstack_objects.py:0:0-0:0) | [formstack/objects.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/formstack/objects.py:0:0-0:0) |
| [test_middleware.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/tests/test_middleware.py:0:0-0:0) | [middleware/espauthmiddleware.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/middleware/espauthmiddleware.py:0:0-0:0) |
| [test_middleware_error.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/tests/test_middleware_error.py:0:0-0:0) | [middleware/esperrormiddleware.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/middleware/esperrormiddleware.py:0:0-0:0) |
| [test_miniblog_models.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/tests/test_miniblog_models.py:0:0-0:0) | [miniblog/models.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/miniblog/models.py:0:0-0:0) |
| [test_program_controllers.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/tests/test_program_controllers.py:0:0-0:0) | [program/controllers/confirmation.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/program/controllers/confirmation.py:0:0-0:0) |
| [test_survey_models.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/tests/test_survey_models.py:0:0-0:0) | `survey/models.py` |
| [test_varnish.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/tests/test_varnish.py:0:0-0:0) | [varnish/varnish.py](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/varnish/varnish.py:0:0-0:0) |

### Documentation Updates

- **README.md** — Added a Testing section with test suite overview, run commands, and PR requirement
- **docs/dev/contributing.rst** — Added a Testing section requiring all tests pass before PR submission

## Running Tests

```bash
# All tests (Docker)
docker exec -it esp-website-web-1 bash -c "cd /code/esp && python manage.py test --settings=esp.settings 2>&1"

# Central test suite only
docker exec -it esp-website-web-1 bash -c "cd /code/esp && python manage.py test esp.tests --settings=esp.settings 2>&1"


### Proposed Solution

Add comprehensive unit tests for all previously untested core modules to ensure stability and prevent regressions.

Specifically, this solution:
1. Adds **16 new test files** covering `accounting`, `application`, [cal](cci:1://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/tests/test_context_processors.py:39:4-47:37), `context_processors`, `db`, `dbmail`, [formstack](cci:1://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/application/models.py:56:4-60:38), `middleware`, `miniblog`, [program](cci:1://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/context_processors.py:22:0-33:13), `survey`, and [varnish](cci:1://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/varnish/varnish.py:5:0-11:19) (adding ~140 new test methods).
2. Fixes several minor edge-case bugs that were uncovered in the models during test creation.
3. Updates project documentation ([README.md](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/README.md:0:0-0:0) and [docs/dev/contributing.rst](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/docs/dev/contributing.rst:0:0-0:0)) to explicitly document the test suite, outline how to run tests via Docker, and mandate that all tests must pass before a Pull Request is merged.


### Alternatives Considered

None. Enforcing comprehensive test coverage and updating the contribution requirements is a standard best practice for maintaining a healthy codebase.


### Additional Context

<details>
<summary><b>Proof of passing tests</b></summary>
The test suite now successfully runs and passes all 171 tests across the project:

```text
test_ordering_newest_first (esp.tests.test_miniblog_models.EntryTest) ... ok
test_str_with_slug (esp.tests.test_miniblog_models.EntryTest) ... ok
test_str_without_slug (esp.tests.test_miniblog_models.EntryTest) ... ok
test_timestamp_auto_set (esp.tests.test_miniblog_models.EntryTest) ... ok
test_no_duplicate_without_repeat (esp.tests.test_program_controllers.ConfirmationEmailControllerTest) ... ok
test_repeat_sends_again (esp.tests.test_program_controllers.ConfirmationEmailControllerTest) ... ok
test_send_confirmation_email_creates_record (esp.tests.test_program_controllers.ConfirmationEmailControllerTest) ... ok
test_answer_property_setter_and_getter (esp.tests.test_survey_models.AnswerTest) ... ok
test_str (esp.tests.test_survey_models.AnswerTest) ... ok
test_get_empty_string (esp.tests.test_survey_models.ListFieldTest) ... ok
test_get_returns_tuple (esp.tests.test_survey_models.ListFieldTest) ... ok
test_set (esp.tests.test_survey_models.ListFieldTest) ... ok
test_get_params (esp.tests.test_survey_models.QuestionTest) ... ok
test_str (esp.tests.test_survey_models.QuestionTest) ... ok
test_is_countable (esp.tests.test_survey_models.QuestionTypeTest) ... ok
test_is_numeric (esp.tests.test_survey_models.QuestionTypeTest) ... ok
test_str_with_params (esp.tests.test_survey_models.QuestionTypeTest) ... ok
test_str (esp.tests.test_survey_models.SurveyResponseTest) ... ok
test_time_filled_auto (esp.tests.test_survey_models.SurveyResponseTest) ... ok
test_num_participants_with_responses (esp.tests.test_survey_models.SurveyTest) ... ok
test_num_participants_zero (esp.tests.test_survey_models.SurveyTest) ... ok
test_str (esp.tests.test_survey_models.SurveyTest) ... ok
test_returns_host_when_configured (esp.tests.test_varnish.GetVarnishHostTest) ... ok
test_returns_none_when_not_configured (esp.tests.test_varnish.GetVarnishHostTest) ... ok
test_returns_none_when_no_varnish (esp.tests.test_varnish.PurgeAllTest) ... ok
test_sends_ban_request (esp.tests.test_varnish.PurgeAllTest) ... ok
test_returns_none_when_no_varnish (esp.tests.test_varnish.PurgePageTest) ... ok
test_sends_purge_request (esp.tests.test_varnish.PurgePageTest) ... ok

----------------------------------------------------------------------
Ran 171 tests in 3.357s

OK

### PR DESCRIPTION
**Pull request by Dhirenderchoudhary**
Date:24 Feb 2026
Developer Name: Dhirender

## Description

This PR adds comprehensive unit tests for all previously untested core modules to ensure stability and prevent regressions. It also updates the project documentation to formalize testing requirements for future contributions.

Specifically, this PR:
1. Adds **16 new test files** covering `accounting`, `application`, [cal](cci:1://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/tests/test_context_processors.py:39:4-47:37), `context_processors`, `db`, `dbmail`, [formstack](cci:1://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/application/models.py:56:4-60:38), `middleware`, `miniblog`, [program](cci:1://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/context_processors.py:22:0-33:13), `survey`, and [varnish](cci:1://file:///Users/nexus/Desktop/ESP-website/ESP-Website/esp/esp/varnish/varnish.py:5:0-11:19) (adding ~140 new test methods).
2. Fixes several minor edge-case bugs that were uncovered in the models during test creation.
3. Updates [README.md](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/README.md:0:0-0:0) and [docs/dev/contributing.rst](cci:7://file:///Users/nexus/Desktop/ESP-website/ESP-Website/docs/dev/contributing.rst:0:0-0:0) to explicitly document the test suite, outline how to run tests via Docker, and mandate that all tests must pass before a Pull Request is merged.

## Related Issue
#4382 

Closes #4382 

## Type of Change

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [x] Documentation update
- [ ] Refactor / cleanup
- [x] CI/CD or build change

## Testing

- [x] Existing tests pass
- [x] New tests added (if applicable)
- [x] Manually verified

## AI Disclosure
 Fully reviewed and verified to pass the suite.

## Checklist

- [x] Self-reviewed the code
- [x] Updated documentation (if needed)
- [x] No new warnings or errors introduced

## Screenshot of all tests passed
Screenshot 2026-02-24 at 8 41 52 PM

### CODE REVIEW

**Comment by Dhirenderchoudhary** [comment]
Hey @willgearty @hwatheod 
 I just opened a new PR to fix the lack of test coverage across the codebase.

Added 16 new test files and ~140 new tests for all the untested modules.
Fixed a few minor model bugs that the new tests uncovered.

Updated 
README.md and contributing.rs with instructions on how to run tests in Docker and a new rule that all PRs must pass tests before merging.
All 171 tests are passing cleanly now. Let me know if you have any feedback or if it's good to merge!

**Comment by willgearty** [comment]
Thanks for putting this together @Dhirenderchoudhary. First, - Please fix lint tests; looks like you have one line with whitespace.

Also, test files should go near their respective views/models. For example, program tests are here: https://github.com/learning-unlimited/ESP-Website/blob/main/esp/esp/program/tests.py. Please move and rename the test files accordingly.

**Comment by Dhirenderchoudhary** [comment]
@willgearty i'll fix it in few minutes sorry for inconvience.

File: docs/dev/contributing.rst [hunk 1]
```diff
@@ -92,6 +92,28 @@ In our workflow, there is generally no need to rebase or squash.  If you are wor
...
+
+Without Docker::
+
+  cd esp
+  python manage.py test --settings=esp.settings
+
+Test files live in ``esp/esp/tests/`` and are named after the source module they
+cover (e.g. ``test_accounting_models.py`` → ``accounting/models.py``). When
+adding new functionality, add corresponding tests following this naming
+convention: ``test_<module>_<component>.py``.
```
Comment (willgearty) [inline docs/dev/contributing.rst]: Need to update this now that most files have been moved.

Comment (Dhirenderchoudhary) [reply docs/dev/contributing.rst]: Updated it

File: docs/dev/contributing.rst [hunk 2]
```diff
@@ -92,6 +92,28 @@ In our workflow, there is generally no need to rebase or squash.  If you are wor
...
+-------
+
+**All tests must pass before submitting a pull request.** You can run the full
+test suite using Docker::
+
+  docker exec -it esp-website-web-1 bash -c "cd /code/esp && python manage.py test --settings=esp.settings 2>&1"
+
+Or run a specific test file::
+
+  docker exec -it esp-website-web-1 bash -c "cd /code/esp && python manage.py test esp.accounting.tests --settings=esp.settings 2>&1"
```
Comment (willgearty) [inline docs/dev/contributing.rst]: The new docker instructions specify this:
```
docker compose exec web python esp/manage.py test
```

File: README.md [hunk 3]
```diff
@@ -17,6 +17,57 @@ docker compose up --build
...
+
+**Without Docker:**
+```bash
+cd esp
+python manage.py test --settings=esp.settings
+```
+
+### ⚠️ PR Requirement
+
+> **All test cases must pass before submitting a Pull Request.** If your changes break existing tests, fix them before requesting a review. If you add new functionality, add corresponding tests in `esp/esp/tests/` following the naming convention `test_<module>_<component>.py`.
```
Comment (willgearty) [inline README.md]: Let's either move this to `contributing.rst` or a new `testing.rst` file.

Comment (Dhirenderchoudhary) [reply README.md]: updated in contributing.rst that before committing all tests must pass.

File: docs/dev/contributing.rst (hunk 2 above)
Comment (Dhirenderchoudhary) [reply docs/dev/contributing.rst]: done from my side
